{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c635d8ff",
   "metadata": {},
   "source": [
    "### To determine the regions of interest in the analysis dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97bfbb24",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import required modules\n",
    "import pybedtools, os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a14995d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_chrom(Chr):\n",
    "    \"\"\" Sort BED file in alpha-numeric order by chromosome.\n",
    "\n",
    "    Arguments\n",
    "    ---------\n",
    "        Chr: Chromosome (i.e. chr1)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "        New: New variable to order autosomal chromosomes as well as sex and mitochondrial chromosomes.\n",
    "    \"\"\"\n",
    "\n",
    "    if Chr:\n",
    "        New = Chr[3:]\n",
    "        if New == 'X':\n",
    "            New = 23\n",
    "        elif New == 'Y':\n",
    "            New = 24\n",
    "        elif New == 'M':\n",
    "            New = 25\n",
    "        elif New == 'N':\n",
    "            New = 26\n",
    "        else:\n",
    "            New = int(New)\n",
    "    else:\n",
    "        New = 0\n",
    "    return New"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d011e679",
   "metadata": {},
   "outputs": [],
   "source": [
    "# There is no CNV, but at least TWO experimental designs have MORE than ONE probes matching exon region --> TRUE NEGATIVE\n",
    "def regions_of_interestv(exonfile, gtexome, arrayid1, arrayid2, arrayid3, out):\n",
    "    #Check that the output path exists, if not create it\n",
    "    if not os.path.exists(out):\n",
    "        os.makedirs(out)\n",
    "        \n",
    "    # Create a BedTool object from the exon file \n",
    "    exon = pybedtools.BedTool(exonfile)\n",
    "    # Create a BedTool object from the ground truth CNVs matching coding exons\n",
    "    gt = pybedtools.BedTool(gtexome)\n",
    "    \n",
    "    #Identify which exons are the exons that are the CNVs of ground truth\n",
    "    gt_exon_isec = exon.intersect(gt, wao=True).intersect(gt, c=True)\n",
    "    gt_exon_ = [(f[0],f[1],f[2],f[3],f[4],f[5],f[9],f[21],f[29]) for f in gt_exon_isec]\n",
    "    gt_exon = pybedtools.BedTool(gt_exon_)                   \n",
    "    \n",
    "    ## Calculate the number of probes from each experimental design matching exon regions\n",
    "    #estd20\n",
    "    exon_probestudy1A = gt_exon.intersect(arrayid1[0], c=True)\n",
    "    exon_probestudy1B = gt_exon.intersect(arrayid1[1], c=True)\n",
    "    exon_probestudy1C = gt_exon.intersect(arrayid1[2], c=True)\n",
    "    exon_probestudy1D = gt_exon.intersect(arrayid1[3], c=True)\n",
    "    cat_probesid1 = exon_probestudy1A.cat(exon_probestudy1B, postmerge=False).cat(\n",
    "        exon_probestudy1C, postmerge=False).cat(exon_probestudy1D, postmerge=False).sort()\n",
    "    exon_probestudy1 = cat_probesid1.groupby(g=[1,2,3,4,5,6,7,8,9], c=10, o=[\"sum\"])\n",
    "    #nstd46\n",
    "    exon_probestudy2 = gt_exon.intersect(arrayid2, c=True)\n",
    "    #estd195\n",
    "    exon_probestudy3 = gt_exon.intersect(arrayid3, c=True)\n",
    "    \n",
    "    # Keep only the field containing the number of probes\n",
    "    num_probe_study1_exon = [x[9] for x in exon_probestudy1]\n",
    "    num_probe_study2_exon = [x[9] for x in exon_probestudy2]\n",
    "    num_probe_study3_exon = [x[9] for x in exon_probestudy3]\n",
    "    \n",
    "    # Create a dataframe that stores all information obtained for each exon interval\n",
    "    data = pybedtools.BedTool.to_dataframe(gt_exon, names=['#chrom', 'start', 'end', 'gene', \n",
    "                                                           'transcript','exon', 'CNV_type', 'confidence_score',\n",
    "                                                           'match_CNV']) \n",
    "\n",
    "    data.insert(loc = 9, column = 'num_probes_id1', value = num_probe_study1_exon)\n",
    "    data.insert(loc = 10, column = 'num_probes_id2', value = num_probe_study2_exon)\n",
    "    data.insert(loc = 11, column = 'num_probes_id3', value = num_probe_study3_exon)\n",
    "         \n",
    "    N = []\n",
    "    P = []\n",
    "    no_capture = []\n",
    "    ROI = []\n",
    "    for index, row in data.iterrows():\n",
    "        chrom = row['#chrom']\n",
    "        start = int(row['start'])\n",
    "        end = int(row['end'])\n",
    "        gene_name = row['gene']\n",
    "        gene_id = row['transcript']\n",
    "        gene_num = row['exon']\n",
    "        cnv_type = row['CNV_type']\n",
    "        conf_score = row['confidence_score']\n",
    "        match_cnv = int(row['match_CNV']) #0 -> not match; num -> match\n",
    "        probes_id1 = int(row['num_probes_id1'])\n",
    "        probes_id2 = int(row['num_probes_id2'])\n",
    "        probes_id3 = int(row['num_probes_id3'])\n",
    "        \n",
    "        #if match_cnv == 0 (negative), check at least two experimental designs have probes matching exon region\n",
    "        #If there are at least two experimental designs, it implies that the CNV could have been characterized; \n",
    "        #therefore, it is negative. If there is insufficient probe, or there is no congruence between studies, \n",
    "        #it is not considered negative. \n",
    "        #Two methods are examined: restrictive and highly restrictive. The highly restrictive is the one analyzed \n",
    "        #(more than 1 probe in at least 2 experimental designs). The restrictive one is not considered in the final work \n",
    "        #(at least 1 probe in at least 1 experimental design).\n",
    "        if match_cnv == 0:\n",
    "            roi = 'negative'\n",
    "            ###RESTRICTIVE:\n",
    "            #if (probes_id1 != 0 and probes_id2 != 0) or (probes_id1 != 0 and probes_id3 != 0) or (probes_id2 != 0 and probes_id3 != 0):\n",
    "            ###HIGH RESTRICTIVE:\n",
    "            if (probes_id1>1 and probes_id2>1) or (probes_id1>1 and probes_id3>1) or (probes_id2>1 and probes_id3>1):\n",
    "                N.append((chrom, start, end, gene_name, gene_id, gene_num, cnv_type, conf_score, match_cnv, \n",
    "                          probes_id1, probes_id2, probes_id3, roi))\n",
    "                ROI.append((chrom, start, end, gene_name, gene_id, gene_num, cnv_type, conf_score, match_cnv, \n",
    "                          probes_id1, probes_id2, probes_id3, roi))\n",
    "            else:\n",
    "                #pass\n",
    "                no_capture.append((chrom, start, end, gene_name, gene_id, gene_num, cnv_type, conf_score, match_cnv, \n",
    "                          probes_id1, probes_id2, probes_id3, roi))\n",
    "        else:\n",
    "            roi = 'positive'\n",
    "            P.append((chrom, start, end, gene_name, gene_id, gene_num, cnv_type, conf_score, match_cnv, \n",
    "                          probes_id1, probes_id2, probes_id3, roi))\n",
    "            ROI.append((chrom, start, end, gene_name, gene_id, gene_num, cnv_type, conf_score, match_cnv, \n",
    "                          probes_id1, probes_id2, probes_id3, roi))\n",
    "            \n",
    "    N_sorted = sorted(N, key=lambda x: (x[0], x[1], x[2]))\n",
    "    P_sorted = sorted(P, key=lambda x: (x[0], x[1], x[2]))\n",
    "    no_capture_sort = sorted(no_capture, key=lambda x: (x[0], x[1], x[2]))\n",
    "    ROI_sorted = sorted(ROI, key=lambda x: (x[0], x[1], x[2]))\n",
    "    \n",
    "    # Write the BED file\n",
    "    with open(os.path.join(out, \"negative.bed\"), 'w') as n:\n",
    "        n.write('#chrom\\tstart\\tend\\tgene\\ttranscript\\texon\\tcnv_type\\tconf_score\\tmatch_cnv\\t\\\n",
    "        probes_id1\\tprobes_id2\\tprobes_id3\\tclass\\n')\n",
    "        n.write(\"\\n\".join(map(lambda x: \"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\" % \n",
    "                              (x[0], x[1], x[2], x[3], x[4], x[5], x[6], x[7], x[8], x[9], x[10], x[11], x[12]), \n",
    "                              N_sorted)))\n",
    "    with open(os.path.join(out, \"negative_no_capture.bed\"), 'w') as nn:\n",
    "        nn.write('#chrom\\tstart\\tend\\tgene\\ttranscript\\texon\\tcnv_type\\tconf_score\\tmatch_cnv\\t\\\n",
    "        probes_id1\\tprobes_id2\\tprobes_id3\\tclass\\n')\n",
    "        nn.write(\"\\n\".join(map(lambda x: \"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\" % \n",
    "                              (x[0], x[1], x[2], x[3], x[4], x[5], x[6], x[7], x[8], x[9], x[10], x[11], x[12]), \n",
    "                      no_capture_sort)))\n",
    "    with open(os.path.join(out, \"positive.bed\"), 'w') as p:\n",
    "        p.write('#chrom\\tstart\\tend\\tgene\\ttranscript\\texon\\tcnv_type\\tconf_score\\tmatch_cnv\\t\\\n",
    "        probes_id1\\tprobes_id2\\tprobes_id3\\tclass\\n')\n",
    "        p.write(\"\\n\".join(map(lambda x: \"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\" % \n",
    "                              (x[0], x[1], x[2], x[3], x[4], x[5], x[6], x[7], x[8], x[9], x[10], x[11], x[12]), \n",
    "                              P_sorted)))\n",
    "    with open(os.path.join(out, \"ROI.bed\"), 'w') as roi:\n",
    "        roi.write('#chrom\\tstart\\tend\\tgene\\ttranscript\\texon\\tcnv_type\\tconf_score\\tmatch_cnv\\t\\\n",
    "        probes_id1\\tprobes_id2\\tprobes_id3\\tclass\\n')\n",
    "        roi.write(\"\\n\".join(map(lambda x: \"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\" % \n",
    "                              (x[0], x[1], x[2], x[3], x[4], x[5], x[6], x[7], x[8], x[9], x[10], x[11], x[12]), \n",
    "                              ROI_sorted)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a003598d",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_of_interest('exome_INGEMM.no_pseudo.bed', 'ground_truth_exome_indet.bed', \n",
    "                    ['nimblegen_10M_A.bed.gz', 'nimblegen_10M_B.bed.gz', \n",
    "                     'nimblegen_10M_C.bed.gz', 'nimblegen_10M_D.bed.gz'],\n",
    "                    ['A-GEOD-11386_hg19.bed', 'A-GEOD-11387_hg19.bed'],\n",
    "                    ['Human1Mv1_C-b37.strand.bed', 'GenomeWideSNP_6.hg19_parsed.bed'], \n",
    "                    'roi_highrestrict')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
